src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_binary.cs

```diff
@@ -22,14 +23,12 @@ public static partial class ReaderExtensions
         public static unsafe bool TryRead<T>(ref this BufferReader<byte> reader, out T value) where T : unmanaged
         {
             ReadOnlySpan<byte> span = reader.UnreadSpan;
-            if (span.Length >= sizeof(T))
-            {
-                value = MemoryMarshal.Read<T>(span);
-                reader.Advance(sizeof(T));
-                return true;
-            }
+            if (span.Length < sizeof(T))
+                return TryReadSlow(ref reader, out value);
 
-            return TryReadSlow(ref reader, out value);
+            value = Unsafe.ReadUnaligned<T>(ref MemoryMarshal.GetReference(span));
```

Comment:
Why is this necessary? I would imagine the redundant length check in MemoryMarshal.Read would be ignored by the JIT.

cc @AndyAyersMS 

https://github.com/dotnet/corefx/blob/71c1371d56fea965c6743bf770c46408dbbdae05/src/Common/src/CoreLib/System/Runtime/InteropServices/MemoryMarshal.cs#L141
```C#
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static T Read<T>(ReadOnlySpan<byte> source)
            where T : struct
        {
            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
            {
                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
            }
            if (Unsafe.SizeOf<T>() > source.Length)
            {
                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length);
            }
            return Unsafe.ReadUnaligned<T>(ref GetReference(source));
        }
```

Reply:
> Why is this necessary?

Don't know precisely, but it makes a huge difference.

Reply:
Other than the redundant length check in the `IG03` block I don't see much difference in the jitted code between the two.
```asm
;; slow version
G_M12736_IG01:
       4883EC38             sub      rsp, 56
       90                   nop      

G_M12736_IG02:
       488D4108             lea      rax, bword ptr [rcx+8]
       4C8B00               mov      r8, bword ptr [rax]
       4C89442430           mov      bword ptr [rsp+30H], r8
       8B4008               mov      eax, dword ptr [rax+8]
       448B4118             mov      r8d, dword ptr [rcx+24]
       443BC0               cmp      r8d, eax
       7774                 ja       SHORT G_M12736_IG10

G_M12736_IG03:
       4C8B4C2430           mov      r9, bword ptr [rsp+30H]
       412BC0               sub      eax, r8d
       4D63C0               movsxd   r8, r8d
       4D03C1               add      r8, r9
       83F804               cmp      eax, 4
       7C56                 jl       SHORT G_M12736_IG08
       83F804               cmp      eax, 4
       7C62                 jl       SHORT G_M12736_IG11

G_M12736_IG04:
       4C89442428           mov      bword ptr [rsp+28H], r8
       488B442428           mov      rax, bword ptr [rsp+28H]
       8B00                 mov      eax, dword ptr [rax]
       8902                 mov      dword ptr [rdx], eax
       488B11               mov      rdx, qword ptr [rcx]
       4883C204             add      rdx, 4
       488911               mov      qword ptr [rcx], rdx
       8B5118               mov      edx, dword ptr [rcx+24]
       4863C2               movsxd   rax, edx
       4C8D4108             lea      r8, bword ptr [rcx+8]
       4D8B08               mov      r9, bword ptr [r8]
       458B4008             mov      r8d, dword ptr [r8+8]
       4D63C0               movsxd   r8, r8d
       4983E804             sub      r8, 4
       4C3BC0               cmp      r8, rax
       7E08                 jle      SHORT G_M12736_IG05
       83C204               add      edx, 4
       895118               mov      dword ptr [rcx+24], edx
       EB0A                 jmp      SHORT G_M12736_IG06

G_M12736_IG05:
       BA04000000           mov      edx, 4
       E83CCAFFFF           call     BufferReader`1:AdvanceToNextSpan(long):this

G_M12736_IG06:
       B801000000           mov      eax, 1

G_M12736_IG07:
       4883C438             add      rsp, 56
       C3                   ret      

G_M12736_IG08:
       E865CAFFFF           call     X:TryReadSlow(byref,byref):bool
       90                   nop      

G_M12736_IG09:
       4883C438             add      rsp, 56
       C3                   ret      

G_M12736_IG10:
       E81A3FE64F           call     ThrowHelper:ThrowArgumentOutOfRangeException()
       CC                   int3     

G_M12736_IG11:
       B919000000           mov      ecx, 25
       E8AF42E64F           call     ThrowHelper:ThrowArgumentOutOfRangeException(int)
       CC                   int3     

```
We really should teach the jit about ref structs . It is being super-paranoid about re-fetching fields. Closest open issue is dotnet/coreclr#19185.

Reply:
> Other than the redundant length check

Presumably, it is this check: `if (Unsafe.SizeOf<T>() > source.Length)`

Can the JIT elide that redundant check if we have already validated that `source.Length >= Unsafe.SizeOf<T>()` before calling the method?

Reply:
Hard to say for sure -- the jit is currently not very good at removing redundant checks in "straight line" code.